src/python/pants/engine/legacy/structs.py

```diff
@@ -310,6 +311,88 @@ def get_sources(self) -> "GlobsWithConjunction":
     return GlobsWithConjunction.for_literal_files(['register.py'], self.address.spec_path)
 
 
+# TODO(#4535): Find a less clunky solution for precise file args as part of the Target API. The
+# trick here is that we must have some way to preserve the OriginSpec while also having a distinct
+# type for each target so that unions work properly.
```

Comment:
As mentioned in the PR description, I do not like the changes to this file at all. It's clunky due to the lack of the target API. But it helps to guide the design for when Stu and I work on this in March!

Reply:
Based on the description (which is great: thanks!), it sounds like we are attempting to project origin information along with addresses in some cases. Rather than being related to #4535, I think that this might be more related to #7490?

What #7490 would allow for would be approximately:
```
.. = await Get[TestResult](Params(Address(..), Origin(..)))
```
...or doing the same thing here to project a `HydratedTarget` without adding a new `HydratedTargetWithOrigin` type.

Moreover, I think that #7490 is probably fairly straightforward, aside from deciding on the syntax... there should be relatively little FFI code, I think. I don't want to suggest blocking this on that... but... you might want to?

Reply:
> it sounds like we are attempting to project origin information along with addresses in some cases

Yes, this is correct. 

> Rather than being related to #4535, I think that this might be more related to #7490?

Interesting!! I like it! I'll update the TODOs to point to this.